firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java

```diff
@@ -190,71 +365,27 @@ private void applyLocalMutationsToDocuments(
       Query query, SnapshotVersion sinceReadTime) {
     ImmutableSortedMap<DocumentKey, MutableDocument> remoteDocuments =
         remoteDocumentCache.getAllDocumentsMatchingQuery(query, sinceReadTime);
+    ImmutableSortedMap<DocumentKey, MutableDocument> localMatches =
+        localDocumentCache.getAllDocumentsMatchingQuery(query);
 
-    List<MutationBatch> matchingBatches = mutationQueue.getAllMutationBatchesAffectingQuery(query);
-
-    remoteDocuments = addMissingBaseDocuments(matchingBatches, remoteDocuments);
-
-    for (MutationBatch batch : matchingBatches) {
-      for (Mutation mutation : batch.getMutations()) {
-        // Only process documents belonging to the collection.
-        if (!query.getPath().isImmediateParentOf(mutation.getKey().getPath())) {
-          continue;
-        }
-
-        DocumentKey key = mutation.getKey();
-        MutableDocument document = remoteDocuments.get(key);
-        if (document == null) {
-          // Create invalid document to apply mutations on top of
-          document = MutableDocument.newInvalidDocument(key);
-          remoteDocuments = remoteDocuments.insert(key, document);
-        }
-        mutation.applyToLocalView(document, batch.getLocalWriteTime());
-        if (!document.isFoundDocument()) {
-          remoteDocuments = remoteDocuments.remove(key);
+    ImmutableSortedMap<DocumentKey, Document> results = emptyDocumentMap();
+    for (Map.Entry<DocumentKey, MutableDocument> localEntry : localMatches) {
+      if (remoteDocuments.containsKey(localEntry.getKey())) {
+        SnapshotVersion version = remoteDocuments.get(localEntry.getKey()).getVersion();
+        if (localEntry.getValue().getVersion().compareTo(version) > 0) {
```

Comment:
Should this not always be true? I assume that we can delete the local entry once the remote has caught up.

Reply:
Yeah, delete when remote caught up is not implemented in prototype.